Add new gall guide. (#1249)
urbit/urbit.org: content/docs/userspace/gall-guide/3-imports-and-aliases.md

use of both libraries.

Let's look at each in more detail:

### `default-agent`

The `default-agent` library contains a basic agent with sane default behaviours
for each arm. In some cases it just crashes and prints an error message to the
terminal, and in others it succeeds but does nothing. It has two primary uses:

- For any agent arms you don't need, you can just have them call the matching
  function in `default-agent`, rather than having to manually handle events on
  those arms.
- A common pattern in an agent is to switch on the input of an arm with
  [wutlus](/docs/hoon/reference/rune/wut#-wutlus) (`?+`) runes or maybe
  [wutcol](/docs/hoon/reference/rune/wut#wutcol) (`?:`) runes. For any
  unexpected input, you can just pass it to the relevant arm of `default-agent`
  rather than handling it manually.

The `default-agent` library lives in `/lib/default-agent/hoon` of the `%base`
desk, and you would typically include a copy in any new desk you created. It's
imported at the beginning of an agent with the
[faslus](/docs/arvo/ford/ford#ford-runes) (`/+`) rune.

The library is a wet gate which takes two arguments: `agent` and `help`. The

> first is your agent core itself, and the second is a `?`. If `help` is `%.y`, it

Reviewer: ```suggestion
first is your agent core itself, and the second is a `?`. If `help` is `%.y` (equivalently, `%&`), it
```
important to mention this other convention since it appears frequently

> will crash in all cases. If `help` is `%.n`, it will use its defaults. You would

Reviewer: ```suggestion
will crash in all cases. If `help` is `%.n` (equivalently, `%|`), it will use its defaults. You would
```

almost always have `help` as `%.n`.

> The wet gate returns a `agent:gall` `door` with a sample of `bowl:gall` - a

Reviewer: ```suggestion
The wet gate returns a `agent:gall` door with a sample of `bowl:gall` - a
```

> The wet gate returns a `agent:gall` door with a sample of `bowl:gall` - a

Reviewer 2: ```suggestion
The wet gate returns an `agent:gall` door with a sample of `bowl:gall` - a
```

typical agent core. Usually you would define an alias for it in a virtual arm
([explained below](#virtual-arms)) so it's simple to call.

### `dbug`

The `dbug` library lets you inspect the state and `bowl` of your agent from the
dojo. It includes an `agent:dbug` function which wraps your whole `agent:gall`

> `door`, adding its extra debugging functionality while transparently passing

Reviewer: ```suggestion
door, adding its extra debugging functionality while transparently passing
```

events to your agent for handling like usual.

To use it, you just import `dbug` with a
[faslus](/docs/arvo/ford/ford#ford-runes) (`/+`) rune at the beginning, then add
the following line directly before the door of your agent:

```hoon
%-  agent:dbug
```

With that done, you can poke your agent with the `+dbug` generator from the dojo
and it will pretty-print its state, like:

```
> :your-agent +dbug
```

The generator also has a few useful optional arguments:

- `%bowl`: Print the agent's bowl.
- `[%state %some-face]`: Print the part of your agent's state that has a face of

>   `some-face`.

Reviewer 2: This interpretation works, but is not fully correct. The actual semantics here are `[%state 'some code']`, where the code is evaluated against a subject which has the app's state at the head. So yes, `someface` will evaluate to the value of someface, but you could also `(~(has by somemap) %index)` etc etc.

- `[%incoming ...]`: Print details of the matching incoming subscription, one
  of:
  - `[%incoming %ship ~some-ship]`
  - `[%incoming %path /part/of/path]`
  - `[%incoming %wire /part/of/wire]`
- `[%outgoing ...]`: Print details of the matching outgoing subscription, one
  of:
  - `[%outgoing %ship ~some-ship]`
  - `[%outgoing %path /part/of/path]`
  - `[%outgoing %wire /part/of/wire]`
  - `[outgoing %term %agent-name]`

By default it will retrieve your agent's state by using its `on-save` arm, but

> if you include a scry endpoint with a path of `/x/dbug/state`, it will use that

Reviewer 2: ```suggestion
By default it will retrieve your agent's state by using its `on-save` arm, but
if your app implements a scry endpoint with a path of `/x/dbug/state`, it will use that
```

instead.

We haven't yet covered some of the concepts described here, so don't worry if

> you don't fully understand `%dbug`'s functionality - you can refer back here

Reviewer: ```suggestion
you don't fully understand `dbug`'s functionality - you can refer back here
```

later.

## Virtual arms

An agent core must have exactly ten arms. However, there's a special kind of
"virtual arm" that can be added without actually increasing the core's arm

> count. A virtual arm is created with the

Reviewer: ```suggestion
count, since it really just adds code to the other arms in the core. A virtual arm is created with the
```

[lustar](/docs/hoon/reference/rune/lus#-lustar) (`+*`) rune, and its purpose is

> to define _aliases_. It takes a list of pairs of names (aliases) and hoon

Reviewer: ```suggestion
to define _deferred expressions_. It takes a list of pairs of names and Hoon
```
See #1254

> expressions. When compiled, the aliases defined in the virtual arm are

Reviewer: ```suggestion
expressions. When compiled, the deferred expressions defined in the virtual arm are
```

implicitly inserted at the beginning of every other arm of the core, so they all

> have access to them.

Reviewer: ```suggestion
have access to them. Each time a name in a `+*` is called, the associated Hoon is evaluated in its place, similar to lazy evaluation except it is re-evaluated whenever needed. See the [tistar](/docs/hoon/reference/rune/tis#-tistar) reference for more information on deferred expressions.
```

> A virtual arm in an agent might look something like this:

Reviewer: ```suggestion
A virtual arm in an agent often looks something like this:
```

```hoon
+*  this  .
    def   ~(. (default-agent this %.n) bowl)
    hc    ~(. +> bowl)
```

> `this`, `def` and `hc` are the aliases, and next to each one is the hoon

Reviewer: ```suggestion
`this`, `def` and `hc` are the deferred expressions, and next to each one is the Hoon
```

> expresion it's aliasing. Notice that unlike most things that take _n_ arguments,

Reviewer: ```suggestion
expression it evaluates whenever called. Notice that unlike most things that take _n_ arguments,
```

a virtual arm is not terminated with a `==`. You can define as many aliases as
you like. The three in this example are conventional ones you'd use in most
agents you write. Their purposes are:

```hoon
this  .
```

Rather than having to return `..on-init` like we did in the last lesson,
instead our arms can just refer to `this` whenever modifying or returning the
agent core.

```hoon
def  ~(. (default-agent this %.n) bowl)
```

This sets up the `default-agent` library we [described above](#default-agent),
so you can easily call its arms like `on-poke:def`, `on-agent:def`, etc.

```hoon
hc  ~(. +> bowl)
```

This makes it easier to call a function in a _helper core_. We'll introduce

> helper cores later, so don't worry about this for now.

Reviewer 2: Consider just not including this example here then. There's already two!

the concepts discussed here:

```hoon
/+  default-agent, dbug
|%
+$  card  card:agent:gall
--
%-  agent:dbug
^-  agent:gall
|_  =bowl:gall
+*  this  .
    def   ~(. (default-agent this %.n) bowl)
++  on-init
  ^-  (quip card _this)
  `this
++  on-save   on-save:def
++  on-load   on-load:def
++  on-poke   on-poke:def
++  on-watch  on-watch:def
++  on-leave  on-leave:def
++  on-peek   on-peek:def
++  on-agent  on-agent:def
++  on-arvo   on-arvo:def
++  on-fail   on-fail:def
--
```

The first line uses the faslus (`/+`) Ford rune to import
`/lib/default-agent.hoon` and `/lib/dbug.hoon`, building them and loading them
into the subject of our agent so they're available for use. You can read more
about Ford runes in the [Ford section of the vane
documenation](/docs/arvo/ford/ford#ford-runes).

Next, we've added an extra core. Notice how it's not explicitly composed, since
the build system will do that for us. In this case we've just added a single
`card` arm, which makes it simpler to reference the `card:agent:gall` type.

After that core, we call `agent:dbug` with our whole agent core as its argument.
This allows us to use the `dbug` features described earlier.

> Inside our agent `door`, we've added an extra virtual arm and defined a couple

Reviewer: ```suggestion
Inside our agent door, we've added an extra virtual arm and defined a couple
```

> of aliases:

Reviewer: ```suggestion
deferred expressions:
```

```hoon
+*  this  .
    def   ~(. (default-agent this %.n) bowl)
```

In most of the arms, you see we've been able to replace the dummy code with
simple calls to the corresponding arms of `default-agent`, which we set up and

> aliased as `def` in the virtual arm. We've also replaced the old `..on-init`

Reviewer: ```suggestion
simple calls to the corresponding arms of `default-agent`, which we set up as a deferred
expression named `def` in the virtual arm. We've also replaced the old `..on-init`
```

> with our alias `this` in the `on-init` arm as an example - it makes things a bit

Reviewer: ```suggestion
with our deferred expression named `this` in the `on-init` arm as an example - it makes things a bit
```

simpler.

You can save the code above in `/app/skeleton.hoon` of your `%base` desk like
before and `|commit %base` in the dojo. Additionally, you can start the agent so
we can try out `dbug`. To start it, run the following in the dojo:

```
> |rein %base [& %skeleton]
```

For details of using the `|rein` generator, see the [Dojo
Tools](/docs/userspace/dist/tools#rein) section of the software distribution
documentation.

Now our agent should be running, so let's try out `dbug`. In the dojo, let's try
poking our agent with the `+dbug` generator:

```
>   ~
> :skeleton +dbug
>=
```

> It just printed out `~` in green. Our dummy `skeleton` agent doesn't have any

Reviewer: This doesn't print out in green for me. I'm running on a fake ship booted from the 1.7 binary today. Can you double check the color?

> documentation.

Reviewer 2: Note that with the latest update, `|start %skeleton` should work as well.

state defined, so it's printing out null as a result. Let's try printing the
`bowl` instead:

```
>   [ [our=~zod src=~zod dap=%skeleton]
  [wex={} sup={}]
  act=5
    eny
  0v209.tg795.bc2e8.uja0d.11eq9.qp3b3.mlttd.gmf09.q7ro3.6unfh.16jiu.m9lh9.6jlt8.4f847.f0qfh.up08t.3h4l2.qm39h.r3qdd.k1r11.bja8l
  now=~2021.11.5..13.28.24..e20e
  byk=[p=~zod q=%base r=[%da p=~2021.11.5..12.02.22..f99b]]
]
> :skeleton +dbug %bowl
>=
```

We'll use `dbug` more throughout the guide, but hopefully you should now have an
idea of its basic usage.

## Summary

The key takeaways are:

> - Libraries can be imported with `/+`.

Reviewer: ```suggestion
- Libraries are imported with `/+`.
```

> - `default-agent` is a library that provides default behaviours for Gall agent

Reviewer: ```suggestion
- `default-agent` is a library that provides default behaviors for Gall agent
```
Pretty sure we use American English conventions throughout the docs

Author: whoops yeah I tried to use american spelling throughout but sometimes I forget

  arms.
- `dbug` is a library that lets you inspect the state and `bowl` of an agent
  from the dojo, with the `+dbug` generator.

> - Convenient aliases for hoon expressions can be defined in a virtual arm with

Reviewer: ```suggestion
- Convenient deferred expressions for Hoon expressions can be defined in a virtual arm with
```

  the [lustar](/docs/hoon/reference/rune/lus#-lustar) (`+*`) rune.

> - `this` is a conventional alias for the agent core itself.

Reviewer: ```suggestion
- `this` is a conventional deferred expression name for the agent core itself.
```

> - `def` is a conventional alias for accessing arms in the `default-agent`

Reviewer: ```suggestion
- `def` is a conventional deferred expression name for accessing arms in the `default-agent`
```
